File: hard/887-super-egg-drop/solution.java
/**
 * 887. Super Egg Drop
 * Time: O(k * m) ~ O(k log n)
 * Space: O(k)
 */
class Solution {
    public int superEggDrop(int k, int n) {
        // dp[j] = max floors resolvable with j eggs at the current move count.
        int[] dp = new int[k + 1];
        int moves = 0;

        while (dp[k] < n) {
            moves++;
            for (int j = k; j >= 1; j--) {
                dp[j] = dp[j] + dp[j - 1] + 1;
            }
        }

        return moves;
    }
}

File: hard/887-super-egg-drop/solution.py
class Solution:
    def superEggDrop(self, k: int, n: int) -> int:
        """
        887. Super Egg Drop
        Time: O(k * m) ~ O(k log n)
        Space: O(k)
        """
        # dp[j] = max floors resolvable with j eggs at the current move count.
        dp = [0] * (k + 1)
        moves = 0

        while dp[k] < n:
            moves += 1
            for j in range(k, 0, -1):
                dp[j] = dp[j] + dp[j - 1] + 1

        return moves
